### PR TITLE
BUILD-920: change default installation namespace

### PR DESCRIPTION
**Pull request by avinal**
## Changes

- by default openshift-builds operator should install in its own namespace
  to enable independent upgrades from other operators
- user can still choose their own namespace while installing

Signed-off-by: Avinal Kumar <avinal@redhat.com>

### CODE REVIEW

**Comment by sayan-biswas** [comment]
Are there any other changes you are expecting in this PR? (since its in draft)

**Comment by avinal** [comment]
Yes, some code changes are required because currently the namespace is a constant in the code and doesn't covers if installed in a different namespace.

File: internal/shipwright/build/build.go
Comment (adambkaplan) [inline internal/shipwright/build/build.go:71]: IIRC the `ShipwrightBuild` object is cluster-scoped, so `GetNamespace()` will always return an empty string.

File: internal/shipwright/build/build_test.go
Comment (adambkaplan) [inline internal/shipwright/build/build_test.go:82]: Ditto in the tests - `ShipwrightBuild` is cluster-scoped

**Review by adambkaplan — CHANGES_REQUESTED** [review]

File: internal/shipwright/build/build.go
Comment (sayan-biswas) [reply internal/shipwright/build/build.go:71]: Yes, `ShipwrightBuild` namespace will be blank. 
You can use the downward api to populate an environment variable or more convenient you can get it from `/var/run/secrets/kubernetes.io/serviceaccount/namespace` inside the container.

File: internal/shipwright/build/build.go
Comment (adambkaplan) [inline internal/shipwright/build/build.go:71]: This function does a disk read in every reconcile for something that we don't expect to change in the container's lifetime. I recommend this is called once in `cmd/main.go` and passed to the reconciler as a parameter/struct variable.

File: internal/shipwright/build/build_test.go
Comment (adambkaplan) [inline internal/shipwright/build/build_test.go:82]: I wouldn't invoke this function here - if this test runs within a Kubernetes container, the current namespace name could be _very_ different. In fact, if this runs within a Tekton task, the value here can change depending on how the task is configured.

Use `.To(Equal("openshift-builds"))` instead.

File: internal/shipwright/build/build_test.go
Comment (adambkaplan) [inline internal/shipwright/build/build_test.go:113]: Ditto - use `.To(Equal("openshift-builds"))` instead.

File: internal/common/predicate.go
Comment (adambkaplan) [inline internal/common/predicate.go:22]: Per the linked documentation:

```
Finally, the default namespace to be used for namespaced API operations is placed in a file at /var/run/secrets/kubernetes.io/serviceaccount/namespace in each container.
```

I _assume_ this is the namespace the pod is deployed in?

File: bundle/manifests/openshift-builds-operator.clusterserviceversion.yaml
Comment (adambkaplan) [inline bundle/manifests/openshift-builds-operator.clusterserviceversion.yaml:50]: :+1:

File: config/manifests/bases/openshift-builds-operator.clusterserviceversion.yaml
Comment (adambkaplan) [inline config/manifests/bases/openshift-builds-operator.clusterserviceversion.yaml:20]: :+1:

**Review by adambkaplan — CHANGES_REQUESTED** [review]
The namespace that we deploy the operands to should be set only once at startup, as this value should not change during the pod's lifecycle.

Controller code should receive this namespace as a struct variable/parameter, and pass it through to relevant code as function arguments/param values.

File: internal/shipwright/build/build.go
Comment (sayan-biswas) [reply internal/shipwright/build/build.go:71]: ++1

File: internal/common/constants.go
Comment (adambkaplan) [inline internal/common/constants.go:22]: Please keep this a constant, as this is our "fallback" value.

File: cmd/main.go
Comment (adambkaplan) [inline cmd/main.go:58]: Yikes! Changing a global variable that is intended to be a constant makes code super hard to follow. This will only lead to thorny, hard to debug / "works on my cluster" bugs.

I also would not invoke `FetchCurrentNamespaceName` in the `init` function - this is reserved for stuff that should work 100% of the time in production. Since this function does a disk read, an I/O failure is possible.

File: cmd/main.go
Comment (adambkaplan) [inline cmd/main.go:120]: I'd simply pass this in here, or do something like

```suggestion
                namespace := common.FetchCurrentNamespace()
                setupLog.Info("operator namespace found", "namespace", namespace)
		Shipwright: shipwrightbuild.New(mgr.GetClient(),  namespace),
```

File: internal/common/predicate.go
Comment (adambkaplan) [inline internal/common/predicate.go:24]: Use the constant for the fallback value.

File: internal/common/predicate.go
Comment (adambkaplan) [inline internal/common/predicate.go:26]: Do we need to worry about line endings or other invalid characters?

File: internal/shipwright/build/build.go
Comment (adambkaplan) [inline internal/shipwright/build/build.go:23]: :+1: yes!

File: internal/common/predicate.go
Comment (adambkaplan) [inline internal/common/predicate.go:21]: :thinking: Since this thing is always running in a Kubernetes pod with a service account + token, I wonder if we should have this return error if we can't read the file.

I think we can merge this as is and discuss once it lands.

File: internal/common/predicate.go
Comment (adambkaplan) [inline internal/common/predicate.go:22]: An area for future improvement - create an "inner" function that captures the rest of the logic, but accepts the file name as a parameter. This makes the code unit-testable outside of Kubernetes.

**Review by adambkaplan — CHANGES_REQUESTED** [review]
Only one major "must change this" regarding the `OpenshiftBuildNamespace` constant. Everything else is nice to haves/areas of future improvement.

File: internal/common/predicate.go
Comment (avinal) [reply internal/common/predicate.go:26]: No, because the file has only one string with no line ending, also we are doing a file read instead of line read and then passing the value to a byte to string converter so that ensure that our string is free of other chars.

File: internal/common/predicate.go
Comment (avinal) [reply internal/common/predicate.go:21]: I think we should have a fallback logic for fetching namespace.

File: README.md
Comment (apoorvajagtap) [inline README.md:96]:
```suggestion
   make catalog-build catalog-push IMAGE_TAG_BASE=quay.io/myusername/rh-openshift-builds/operator
```
I think, `catalog-build` has been missed? cc @sayan-biswas

File: README.md
Comment (apoorvajagtap) [inline README.md:80]:
```suggestion
> make docker-build docker-push bundle bundle-build bundle-push catalog-fbc-build catalog-build catalog-push catalog-deploy IMAGE_TAG_BASE=quay.io/musername/rh-openshift-builds/operator
```
same.

**Review by apoorvajagtap — CHANGES_REQUESTED** [review]

File: internal/common/constants.go
Comment (adambkaplan) [inline internal/common/constants.go:23]: I still don't fully understand why we are storing this in a global state.

File: internal/common/predicate.go
Comment (adambkaplan) [inline internal/common/predicate.go:28]: Avoid global state/variables here. Suggestion also includes the typo fix.

```suggestion
	if err != nil {
		return OpenShiftBuildNamespaceName
	}
	return string(namespace)
```

**Review by adambkaplan — APPROVED** [review]
Not blocking merge on this - items that have been identified are minor things that can be addressed in a follow-up pull request.

**Pull request merged**
